build-tools/packages/build-infrastructure/src/filter.ts

```diff
@@ -0,0 +1,296 @@
+			throw new Error(`Can't find a remote with ${fluidRepo.upstreamRemotePartialUrl}`);
+		}
+		const { packages } = await getChangedSinceRef(
+			fluidRepo,
+			selection.changedSinceBranch,
+			remote,
+		);
+		addAllToSet(selected, packages);
+	}
+
+	if (selection.directory !== undefined) {
+		const repoRelativePath = path.join(
+			selection.directory === "." ? process.cwd() : selection.directory,
+		);
+
+		const dirPackage = [...fluidRepo.packages.values()].find(
+			(p) => fluidRepo.relativeToRepo(p.directory) === repoRelativePath,
+		);
+		if (dirPackage === undefined) {
+			throw new Error(`Cannot find package with directory: ${repoRelativePath}`);
+		}
+		selected.add(dirPackage);
+		return selected;
+	}
+
+	// Select workspace and workspace root packages
+	for (const workspace of fluidRepo.workspaces.values()) {
+		if (selection.workspaces.length > 0 && mm.isMatch(workspace.name, selection.workspaces)) {
+			addAllToSet(
+				selected,
+				workspace.packages.filter((p) => !p.isWorkspaceRoot),
+			);
+		}
+
+		if (
+			selection.workspaceRoots.length > 0 &&
+			mm.isMatch(workspace.name, selection.workspaceRoots)
+		) {
+			addAllToSet(
+				selected,
+				workspace.packages.filter((p) => p.isWorkspaceRoot),
+			);
+		}
+	}
+
+	// Select release group and release group root packages
+	for (const releaseGroup of fluidRepo.releaseGroups.values()) {
+		if (
+			selection.releaseGroups.length > 0 &&
+			mm.isMatch(releaseGroup.name, selection.releaseGroups)
+		) {
+			addAllToSet(
+				selected,
+				releaseGroup.packages.filter((p) => !p.isReleaseGroupRoot),
+			);
+		}
+
+		if (
+			selection.releaseGroupRoots.length > 0 &&
+			mm.isMatch(releaseGroup.name, selection.releaseGroupRoots)
+		) {
+			addAllToSet(
+				selected,
+				releaseGroup.packages.filter((p) => p.isReleaseGroupRoot),
+			);
+		}
+	}
+
+	return selected;
+};
+
+/**
+ * Selects packages from the Fluid repo based on the selection criteria. The selected packages will be filtered by the
+ * filter criteria if provided.
+ *
+ * @param fluidRepo - The Fluid repo.
+ * @param selection - The selection criteria to use to select packages.
+ * @param filter - An optional filter criteria to filter selected packages by.
+ * @returns An object containing the selected packages and the filtered packages.
+ */
+export async function selectAndFilterPackages(
+	fluidRepo: IFluidRepo,
+	selection: PackageSelectionCriteria,
+	filter?: PackageFilterOptions,
+): Promise<{ selected: IPackage[]; filtered: IPackage[] }> {
+	// Select the packages from the repo
+	const selected = [...(await selectPackagesFromRepo(fluidRepo, selection))];
+
+	// Filter resulting list if needed
+	const filtered = filter === undefined ? selected : await filterPackages(selected, filter);
+
+	return { selected, filtered };
+}
+
+/**
+ * Convenience type that extracts only the properties of a package that are needed for filtering.
+ */
+type FilterablePackage = Pick<IPackage, "name" | "private">;
```

Comment:
Problem with this approach is that the package name's type is PackageName, and that hurts the usability of this type as a generic.

Reply:
```ts
interface FilterablePackage {
  name: string;
  private?: boolean | undefined;
}
```

Is that better?